[ethernet] Add LAN8670 PHY support

xorbit opened the pull request:
# What does this implement/fix?

This adds support for the LAN867x 10BASE-T1S Ethernet PHY.
Espressif provides support for this PHY in the ESP Component Registry.

This PHY is required to support the [ManT1S board](https://www.crowdsupply.com/silicognition/mant1s).

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#5400

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# Example config.yaml for ManT1S
ethernet:
  type: LAN8670
  mdc_pin:
    number: GPIO8
    ignore_pin_validation_error: true
  mdio_pin:
    number: GPIO7
    ignore_pin_validation_error: true
  clk:
    pin:
      number: GPIO0
      ignore_strapping_warning: true
    mode: CLK_EXT_IN
  phy_addr: 0
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

xorbit committed:
[ethernet] Add LAN867x support

This adds support for the LAN867x 10BASE-T1S Ethernet PHY.
Espressif provides support for this PHY in the ESP
Component Registry.

This PHY is required to support the
[ManT1S board](https://www.crowdsupply.com/silicognition/mant1s).

bdraco commented on esphome/components/ethernet/__init__.py:
```diff
@@ -75,6 +79,7 @@
     "DM9051": EthernetType.ETHERNET_TYPE_DM9051,
+    "LAN8670": EthernetType.ETHERNET_TYPE_LAN8670,
 }
 
```
Probably need to add LAN8670 to `_PHY_TYPE_TO_DEFINE` and than wrap some code with  `USE_ETHERNET_LAN8670`

You'll also need to add compile tests

xorbit replied:
Ok I can do that.

I currently have all LAN867x code wrapped in
```c
#if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(5, 3, 0) && \
    (defined(USE_ESP32_VARIANT_ESP32) || defined(USE_ESP32_VARIANT_ESP32P4))
#endif
```
guards.  Do you suggest I keep those and add `USE_ETHERNET_LAN8670` to the guards?  Or are the other guards superfluous due to checks at other levels and I should just replace my guards with
```c
#ifdef USE_ETHERNET_LAN8670
#endif
```
guards?

Can you point me to where to define compile tests?

bdraco replied:
`The #ifdef USE_ETHERNET_LAN8670 is likely the best option`

Than you can do the variant check in something like `_is_framework_spi_polling_mode_supported` in the python code

xorbit replied:
Looking into this more, it seems to me that the validator currently completely ignores whether ANY of the RMII PHYs work with the current variant.  My variant check was only to make sure the LAN8670 was being used with a chip that actually has RMII, but it seems that the validator will happily accept the config of any RMII PHY with a chip that doesn't have RMII and then throw a compile error.

So I'm going to add a generic variant check for all RMII PHYs, and a version check for LAN8670 support.

xorbit replied:
I was trying to style my ESP-IDF version check after what I see in `_is_framework_spi_polling_mode_supported`, but this doesn't seem to be working correctly when framework type is switched to `arduino`.

Despite the note in `_is_framework_spi_polling_mode_supported` that claims "Arduino now uses ESP-IDF as a component, so we only check IDF version", `CORE.data[KEY_CORE][KEY_FRAMEWORK_VERSION]` seems to return the Arduino version when framework type is `arduino`.

To make sure I actually tested such an SPI config, and switching to framework type `arduino` triggered the error `In this version of the framework (arduino 3.2.1), 'interrupt_pin' is a required option for [ethernet].`  This shows that the current validation for this is not working correctly.

So the bottom line is `_is_framework_spi_polling_mode_supported` is not working correctly as it is and I don't know what code to use to make my `_is_lan867x_component_supported` work correctly either.  What is the correct variable to check to see the ESP-IDF version, even when framework type is `arduino`?

bdraco replied:
Added the validator & missing define (used for IDEs) in 053af22394f7fd9c27cd6db64bc1f3177960dcbc

bdraco commented on esphome/components/ethernet/__init__.py:
```diff
@@ -356,5 +362,9 @@ async def to_code(config):
     # Also disable WiFi/BT coexistence since WiFi is disabled
     add_idf_sdkconfig_option("CONFIG_SW_COEXIST_ENABLE", False)
 
+    if CORE.using_esp_idf:
+        # Add LAN867x 10BASE-T1S PHY support component
+        add_idf_component(name="espressif/lan867x", ref="2.0.0")
```
only add this if its configured.

Also it will work on all ESP32

xorbit replied:
You're saying IDF components work on the Arduino core as well, so the `if CORE.using_esp_idf:` is not needed?

So what should this if statement be to check whether it's configured?  I'm not finding any equivalent code for the other optional PHYs (KSZ8081).

bdraco replied:
Arduino is installed as an IDF component on dev

xorbit replied:
I think I figured it out.

xorbit committed:
[ethernet] Fix LAN8670 PHY support based on review feedback

- Add LAN8670 to _PHY_TYPE_TO_DEFINE
- Only add IDF component if LAN8670 PHY type selected
- Switch to using USE_ETHERNET_LAN8670 guards in .cpp code
- Add compile test YAML

bdraco commented:
Marked as draft to take it out of review queue. 

Please mark ready once https://github.com/esphome/esphome/pull/10874/files#r2377465389 is addressed

bdraco committed:
add validator

bdraco reviewed (dismissed):
Thanks @xorbit

bdraco committed:
Remove USE_ETHERNET_LAN8670 from static analysis defines

The LAN8670 PHY header is provided by an external IDF component that's only
available during actual compilation, not during static analysis. Since
esphome/core/defines.h is only used for development and static analysis tools,
removing this define prevents clang-tidy errors while not affecting runtime.

xorbit commented:
The change you made is fine for the LAN8670 support, but like I said, it might be good to check this is general for all RMII connected PHYs.  This is not specific to the LAN8670.

I was going to add this code:

```python
def _is_rmii_supported():
    # RMII PHYs only work on chips with an RMII peripheral
    variant = get_esp32_variant()
    return variant in (VARIANT_ESP32, VARIANT_ESP32P4)
```

And in the `elif config[CONF_TYPE] != "OPENETH":` block, not in a specific LAN8670 block:

```python
        if not _is_rmii_supported():
            raise cv.Invalid(
                f"The configured PHY {config[CONF_TYPE]} requires an RMII interface and "
                f"the {get_esp32_variant()} variant does not have this interface."
            )
```

xorbit commented:
> Would be good to do in a followup PR that is focus on other devices and not just LAN8670

Ok, new PR added for this #10909.
